rfcs/20240325-reorganize-gpu/README.md

```diff
@@ -0,0 +1,377 @@
+sub-group sizes).
+
+The `dnnl::impl::gpu::engine_t` is defined as follows:
+
+```cpp
+// Location: src/gpu
+namespace dnnl::impl::gpu {
+
+struct engine_t : public dnnl::impl::engine_t {
+    // Returns a pointer to the engine implementation.
+    const dnnl::impl::engine_impl_t *impl() const;
+
+    // Points to either dnnl::impl::sycl::engine_impl_t or dnnl::impl::intel::ocl::engine_impl_t
+    std::unique_ptr<dnnl::impl::engine_impl_t> impl;
+};
+```
+
+The following vendor specific engine classes will replace the currently
+implemented ones.
+
+This class takes over responsibility of `compute_engine_t` class.
+```cpp
+// Location: src/gpu/intel
+namespace dnnl::impl::gpu::intel {
+
+struct engine_t : public dnnl::impl::gpu::engine_t {};
+
+} // namespace dnnl::impl::gpu::intel
+```
+
+This class takes over responsibility of `sycl_engine_base_t` class.
+```cpp
+// Location: src/gpu/intel/sycl
+namespace dnnl::impl::gpu::intel::sycl {
+
+struct engine_t : public dnnl::impl::gpu::intel::engine_t {};
+
+} // namespace dnnl::impl::gpu::intel::sycl
+```
+
+This class takes over responsibility of `ocl_gpu_engine_t` class.
+```cpp
+// Location: src/gpu/intel/ocl
+namespace dnnl::impl::gpu::intel::ocl {
+
+struct engine_t : public dnnl::impl::gpu::intel::engine_t {};
+
+} // namespace dnnl::impl::gpu::intel::ocl
+```
+
+This class takes over responsibility of `sycl_cuda_engine_t` and `sycl_hip_engine_t` classes.
+```cpp
+// Location: src/gpu/nvidia
+namespace dnnl::impl::gpu::nvidia {
+
+struct engine_t : public dnnl::impl::gpu::engine_t {};
+
+} // namespace dnnl::impl::gpu::nvidia
+
+
+// Location: src/gpu/amd
+namespace dnnl::impl::gpu::amd {
+
+struct engine_t : public dnnl::impl::gpu::engine_t {};
+
+} // namespace dnnl::impl::gpu::amd
+```
+
+This class will be used when only generic SYCL kernels are enabled.
+```cpp
+// Location: src/gpu/generic
+namespace dnnl::impl::gpu::generic {
+
+struct engine_t : public dnnl::impl::gpu::engine_t {};
+
+} // namespace dnnl::impl::gpu::generic
+
+```
+
+The present inheritance chains for SYCL and OpenCL GPU engines are
+the following:
+* SYCL Intel and generic: `engine_t` -> `compute_engine_t` -> `sycl_engine_base_t` -> `sycl_gpu_engine_t`
+* SYCL NVIDIA: `engine_t` -> `compute_engine_t` -> `sycl_engine_base_t` -> `sycl_cuda_engine_t`
+* SYCL AMD: `engine_t` -> `compute_engine_t` -> `sycl_engine_base_t` -> `sycl_hip_engine_t`
+* OpenCL (only Intel): `engine_t` -> `compute_engine_t` -> `ocl_gpu_engine_t`
+
+The new inheritance chains for SYCL and OpenCL GPU engines are
+the following:
+* SYCL Generic: `engine_t` -> `gpu::engine_t` -> `gpu::generic::engine_t`
+* SYCL Intel: `engine_t` -> `gpu::engine_t` -> `gpu::intel::engine_t` -> `gpu::intel::sycl::engine_t`
+* SYCL NVIDIA: `engine_t` -> `gpu::engine_t` -> `gpu::nvidia::engine_t`
+* SYCL AMD: `engine_t` -> `gpu::engine_t` -> `gpu::amd::engine_t`
+* OpenCL: `engine_t` -> `gpu::engine_t` ->  `gpu::intel::engine_t` -> `gpu::intel::ocl::engine_t`
+
```

Comment:
Could you elaborate on the plan to share generic implementation lists across sycl engines? 
Would we keep separate implementation list for each vendor (and add generic impls in each of those), make the generic impl list common somehow? ...

Reply:
I added this information to the RFC.